README.rst
LunarCalendar: A Lunar-Solar Converter
======================================

.. image::
  https://img.shields.io/pypi/v/LunarCalendar.svg
  :target: https://pypi.python.org/pypi/LunarCalendar
  :alt: Last stable version (PyPI)

.. image::
  https://travis-ci.org/wolfhong/LunarCalendar.svg
  :target: https://travis-ci.org/wolfhong/LunarCalendar
  :alt: build status


Overview
--------

`Chinese version(中文版) <https://github.com/wolfhong/LunarCalendar/blob/develop/README_zh.rst>`_ is provided.

LunarCalendar is a Lunar-Solar Converter, containing a number of lunar and solar holidays in China.

Korean, Japanese and Vietnamese lunar calendar is the same as Chinese calendar, but has different holidays.
If `LunarCalendar` is extended; Korean, Japanese and Vietnamese holidays are easily included, with their languages.

LunarCalendar supports the time range 1900-2100. But if you have a need for the time range, you can use ``generate.html`` to extend it.

LunarCalendar is inspired by `Lunar-Solar-Calendar-Converter <https://github.com/isee15/Lunar-Solar-Calendar-Converter>`_.


Features
--------

* Accurate raw data, synchronize with Microsolf's ``ChineseLunisolarCalendar`` class
* Easy to extend holidays and languages, supported both `zh_hans` and `zh_hant`
* Included Lunar Festivals, such as: MidAutumn Festival, Chinese New Year Eve, DragonBoat Festivals
* Included Solar Festivals without fixed dates, such as: Mother's Day, Easter
* Added legality check of the lunar and solar date
* Supported 24 solar terms


Install
-------

LunarCalendar can be installed from the PyPI with `easy_install`::

   $ easy_install LunarCalendar

Or pip::

   $ pip install LunarCalendar


Console Commands
----------------

A console command called `lunar-find` can be used to find the date of the festival, using it's chinese name.
Default to this year. Supporting alias of the festival.

.. code-block:: console

    $ lunar-find 重阳
    重阳节 on 2018: 2018-10-17

    $ lunar-find 重陽節
    重阳节 on 2018: 2018-10-17

    $ lunar-find 登高节 2019
    重阳节 on 2019: 2019-10-07

You can also print all included festivals or 24 solar terms by date asc with:

.. code-block:: console

    $ lunar-find all 2019
    $ lunar-find festival 2012
    $ lunar-find solarterm


Quickstart
----------

Solar to Lunar:

.. code-block:: python

    import datetime
    from lunarcalendar import Converter, Solar, Lunar, DateNotExist

    solar = Solar(2018, 1, 1)
    print(solar)
    lunar = Converter.Solar2Lunar(solar)
    print(lunar)
    solar = Converter.Lunar2Solar(lunar)
    print(solar)
    print(solar.to_date(), type(solar.to_date()))

Lunar to Solar:

.. code-block:: python

    lunar = Lunar(2018, 2, 30, isleap=False)
    print(lunar)
    solar = Converter.Lunar2Solar(lunar)
    print(solar)
    lunar = Converter.Solar2Lunar(solar)
    print(lunar)
    print(lunar.to_date(), type(lunar.to_date()))
    print(Lunar.from_date(datetime.date(2018, 4, 15)))

Legality check for solar and lunar date. 2018-2-15(Leap Month) does not exist, but 2012-4-4(Leap Month) exists:

.. code-block:: python

    Lunar(2012, 4, 4, isleap=True)  # date(2012, 5, 24)
    try:
        lunar = Lunar(2018, 2, 15, isleap=True)
    except DateNotExist:
        print(traceback.format_exc())

Print all the festivals included, with Chinese and English. Other languages are welcome to extend(Fork & Pull Request).

.. code-block:: python

    from lunarcalendar.festival import festivals

    # print festivals, using English or Chinese
    print("----- print all festivals on 2018 in chinese: -----")
    for fest in festivals:
        print(fest.get_lang('zh'), fest(2018))

    print("----- print all festivals on 2017 in english: -----")
    for fest in festivals:
        print(fest.get_lang('en'), fest(2017))

Output:

.. code-block:: shell

    ......
    母亲节 2018-05-13
    父亲节 2018-06-17
    中秋节 2018-09-24
    感恩节 2018-11-22
    重阳节 2018-10-17
    春节 2018-02-16
    中元节 2018-08-25
    七夕节 2018-08-17
    腊八节 2019-01-13
    清明节 2018-04-05
    除夕 2019-02-04
    寒衣节 2018-11-08
    元宵节 2018-03-02
    龙抬头 2018-03-18
    端午节 2018-06-18
    ......


Contribution
------------

Including festival standards:

* Common holidays in the the country, such as: Christmas, Halloween, etc.
* Lunar holidays.
* Solar holidays without fixed dates, such as: Mother's Day, Easter, etc.

Supporting Chinese and English only now. If you want to add Korean, Japanese or Vietnamese supports, modify ``lunarcalendar/festival.py`` to add holidays and languages.

Some unusual holidays may not be included, `welcom to extend <https://github.com/wolfhong/LunarCalendar/issues>`_.



About
-----

* `Homepage <http://github.com/wolfhong/LunarCalendar>`_
* `PyPI <https://pypi.python.org/pypi/LunarCalendar>`_
* `Issue tracker <https://github.com/wolfhong/LunarCalendar/issues?status=new&status=open>`_


CHANGELOG
v0.0.8
------

* Supported `all`, `solarterm`, `festival` in `lunar-find` console command
* Added 24 solar terms, #1
* Fixed #2, according to the correct definition of Ching Ming Festival
* Supported both zh-hans and zh-hant, #6
* Added Easter(复活节), PoWu Festival(破五节), XiaoNian Festival(小年)
* Fixed ohter issues


tests/test_calc.py
# -*- coding: utf-8 -*-
# online Lunar Calendar: https://wannianrili.51240.com/
import datetime
from lunarcalendar._calc import specified_solar_term, annual_solar_terms


def test_calc():
    for year in range(1990, 2101):
        annual_solar_terms(year)
        for i in range(0, 24):
            if i < 22:
                assert specified_solar_term(year, i).year == year
            else:
                # minor cold(小寒) | major cold(大寒)
                assert specified_solar_term(year, i).year == year + 1


tests/test_command.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals
from subprocess import Popen, PIPE
from lunarcalendar import zh_festivals, zh_solarterms


def _assert_output_line_count(args, count):
    with Popen(['python', 'lunarcalendar/command.py'] + args, stdout=PIPE) as p:
        assert len([li for li in p.stdout.readlines() if li.strip()]) == count


def test_command():
    _assert_output_line_count(['重阳节'], 1)
    _assert_output_line_count(['重阳节', '2010'], 1)
    _assert_output_line_count(['重阳'], 1)
    _assert_output_line_count(['重阳', '2011'], 1)

    _assert_output_line_count(['登高节'], 1)
    _assert_output_line_count(['登高节', '2012'], 1)
    _assert_output_line_count(['登高節'], 1)
    _assert_output_line_count(['登高節', '2013'], 1)

    _assert_output_line_count(['没有节'], 0)
    _assert_output_line_count(['没有节', '2014'], 0)

    _assert_output_line_count(['all'], len(zh_festivals) + len(zh_solarterms))
    _assert_output_line_count(['all', '2015'], len(zh_festivals) + len(zh_solarterms))

    _assert_output_line_count(['festival'], len(zh_festivals))
    _assert_output_line_count(['festivals', '2016'], len(zh_festivals))

    _assert_output_line_count(['solarterm'], len(zh_solarterms))
    _assert_output_line_count(['solarterms', '2017'], len(zh_solarterms))


tests/test_festival.py
# -*- coding: utf-8 -*-
# online Lunar Calendar: https://wannianrili.51240.com/
import datetime
from lunarcalendar import festival


def test_festivals():
    assert festival.NewYear(2018) == datetime.date(2018, 1, 1)
    assert festival.Valentine(2018) == datetime.date(2018, 2, 14)
    assert festival.WomenDay(2018) == datetime.date(2018, 3, 8)
    assert festival.ArborDay(2018) == datetime.date(2018, 3, 12)
    assert festival.ChingMing(2018) == datetime.date(2018, 4, 5)
    assert festival.ChingMing(2012) == datetime.date(2012, 4, 4)
    assert festival.ChingMing(2088) == datetime.date(2088, 4, 4)

    assert festival.LabourDay(2018) == datetime.date(2018, 5, 1)
    assert festival.YouthDay(2018) == datetime.date(2018, 5, 4)
    assert festival.NurseDay(2018) == datetime.date(2018, 5, 12)
    assert festival.MotherDay(2018) == datetime.date(2018, 5, 13)  # 2th sunday in May
    assert festival.ChildrenDay(2018) == datetime.date(2018, 6, 1)

    assert festival.FatherDay(2018) == datetime.date(2018, 6, 17)  # 3th sunday in June
    assert festival.TeacherDay(2018) == datetime.date(2018, 9, 10)
    assert festival.NationDay(2018) == datetime.date(2018, 10, 1)
    assert festival.Halloween(2018) == datetime.date(2018, 10, 31)
    assert festival.Thanksgiving(2018) == datetime.date(2018, 11, 22)

    assert festival.ChristmasEve(2018) == datetime.date(2018, 12, 24)
    assert festival.ChristmasDay(2018) == datetime.date(2018, 12, 25)
    assert festival.Easter(2019) == datetime.date(2019, 4, 21)

    assert festival.LaBa(2017) == datetime.date(2018, 1, 24)
    assert festival.NewYearEve(2017) == datetime.date(2018, 2, 15)
    assert festival.ChineseNewYear(2018) == datetime.date(2018, 2, 16)
    assert festival.Lantern(2018) == datetime.date(2018, 3, 2)
    assert festival.DragonHead(2018) == datetime.date(2018, 3, 18)

    assert festival.DragonBoat(2018) == datetime.date(2018, 6, 18)
    assert festival.Qixi(2018) == datetime.date(2018, 8, 17)
    assert festival.Ghost(2018) == datetime.date(2018, 8, 25)
    assert festival.MidAutumn(2018) == datetime.date(2018, 9, 24)
    assert festival.ChongYang(2018) == datetime.date(2018, 10, 17)

    assert festival.Hanyi(2018) == datetime.date(2018, 11, 8)
    assert festival.DongJie(2018) == datetime.date(2018, 12, 22)


lunarcalendar/_calc.py
# -*- coding: utf-8 -*-
# rewrite from https://github.com/kiyota-yoji/EastAsiaCalendars
from __future__ import unicode_literals
import ephem
from math import pi, floor, fabs, sin, cos
import pytz


class Data:

    NUT_SCALE = 1e4
    NUT_SERIES = 106
    NUT_MAXMUL = 4
    SECPERCIRC = (3600. * 360.)

    '''Delaunay arguments, in arc seconds; they differ slightly from ELP82B'''
    delaunay = [
        [485866.733,  1717915922.633, 31.310,  0.064],  # M', moon mean anom
        [1287099.804, 129596581.224,  -0.577, -0.012],  # M, sun mean anom
        [335778.877,  1739527263.137, -13.257, 0.011],  # F, moon arg lat
        [1072261.307, 1602961601.328, -6.891,  0.019],  # D, elong moon sun
        [450160.280,  -6962890.539,   7.455,   0.008],  # Om, moon l asc node
    ]

    '''multipliers for Delaunay arguments'''
    # bounds:  -2..3, -2..2, -2/0/2/4, -4..4, 0..2
    multarg = [
        [0, 0, 0, 0, 1],
        [0, 0, 0, 0, 2],
        [-2, 0, 2, 0, 1],
        [2, 0, -2, 0, 0],
        [-2, 0, 2, 0, 2],
        [1, -1, 0, -1, 0],
        [0, -2, 2, -2, 1],
        [2, 0, -2, 0, 1],
        [0, 0, 2, -2, 2],
        [0, 1, 0, 0, 0],
        [0, 1, 2, -2, 2],
        [0, -1, 2, -2, 2],
        [0, 0, 2, -2, 1],
        [2, 0, 0, -2, 0],
        [0, 0, 2, -2, 0],
        [0, 2, 0, 0, 0],
        [0, 1, 0, 0, 1],
        [0, 2, 2, -2, 2],
        [0, -1, 0, 0, 1],
        [-2, 0, 0, 2, 1],
        [0, -1, 2, -2, 1],
        [2, 0, 0, -2, 1],
        [0, 1, 2, -2, 1],
        [1, 0, 0, -1, 0],
        [2, 1, 0, -2, 0],
        [0, 0, -2, 2, 1],
        [0, 1, -2, 2, 0],
        [0, 1, 0, 0, 2],
        [-1, 0, 0, 1, 1],
        [0, 1, 2, -2, 0],
        [0, 0, 2, 0, 2],
        [1, 0, 0, 0, 0],
        [0, 0, 2, 0, 1],
        [1, 0, 2, 0, 2],
        [1, 0, 0, -2, 0],
        [-1, 0, 2, 0, 2],
        [0, 0, 0, 2, 0],
        [1, 0, 0, 0, 1],
        [-1, 0, 0, 0, 1],
        [-1, 0, 2, 2, 2],
        [1, 0, 2, 0, 1],
        [0, 0, 2, 2, 2],
        [2, 0, 0, 0, 0],
        [1, 0, 2, -2, 2],
        [2, 0, 2, 0, 2],
        [0, 0, 2, 0, 0],
        [-1, 0, 2, 0, 1],
        [-1, 0, 0, 2, 1],
        [1, 0, 0, -2, 1],
        [-1, 0, 2, 2, 1],
        [1, 1, 0, -2, 0],
        [0, 1, 2, 0, 2],
        [0, -1, 2, 0, 2],
        [1, 0, 2, 2, 2],
        [1, 0, 0, 2, 0],
        [2, 0, 2, -2, 2],
        [0, 0, 0, 2, 1],
        [0, 0, 2, 2, 1],
        [1, 0, 2, -2, 1],
        [0, 0, 0, -2, 1],
        [1, -1, 0, 0, 0],
        [2, 0, 2, 0, 1],
        [0, 1, 0, -2, 0],
        [1, 0, -2, 0, 0],
        [0, 0, 0, 1, 0],
        [1, 1, 0, 0, 0],
        [1, 0, 2, 0, 0],
        [1, -1, 2, 0, 2],
        [-1, -1, 2, 2, 2],
        [-2, 0, 0, 0, 1],
        [3, 0, 2, 0, 2],
        [0, -1, 2, 2, 2],
        [1, 1, 2, 0, 2],
        [-1, 0, 2, -2, 1],
        [2, 0, 0, 0, 1],
        [1, 0, 0, 0, 2],
        [3, 0, 0, 0, 0],
        [0, 0, 2, 1, 2],
        [-1, 0, 0, 0, 2],
        [1, 0, 0, -4, 0],
        [-2, 0, 2, 2, 2],
        [-1, 0, 2, 4, 2],
        [2, 0, 0, -4, 0],
        [1, 1, 2, -2, 2],
        [1, 0, 2, 2, 1],
        [-2, 0, 2, 4, 2],
        [-1, 0, 4, 0, 2],
        [1, -1, 0, -2, 0],
        [2, 0, 2, -2, 1],
        [2, 0, 2, 2, 2],
        [1, 0, 0, 2, 1],
        [0, 0, 4, -2, 2],
        [3, 0, 2, -2, 2],
        [1, 0, 2, -2, 0],
        [0, 1, 2, 0, 1],
        [-1, -1, 0, 2, 1],
        [0, 0, -2, 0, 1],
        [0, 0, 2, -1, 2],
        [0, 1, 0, 2, 0],
        [1, 0, -2, -2, 0],
        [0, -1, 2, 0, 1],
        [1, 1, 0, -2, 1],
        [1, 0, -2, 2, 0],
        [2, 0, 0, 2, 0],
        [0, 0, 2, 4, 2],
        [0, 1, 0, 1, 0]
    ]

    '''amplitudes which  have secular terms; in 1/NUT_SCALE arc seconds
    {index, constant dPSI, T/10 in dPSI, constant in dEPS, T/10 in dEPS}'''
    ampsecul = [
        [0  ,-171996 ,-1742 ,92025 ,89],
        [1  ,2062    ,2     ,-895  ,5],
        [8  ,-13187  ,-16   ,5736  ,-31],
        [9  ,1426    ,-34   ,54    ,-1],
        [10 ,-517    ,12    ,224   ,-6],
        [11 ,217     ,-5    ,-95   ,3],
        [12 ,129     ,1     ,-70   ,0],
        [15 ,17      ,-1    ,0     ,0],
        [17 ,-16     ,1     ,7     ,0],
        [30 ,-2274   ,-2    ,977   ,-5],
        [31 ,712     ,1     ,-7    ,0],
        [32 ,-386    ,-4    ,200   ,0],
        [33 ,-301    ,0     ,129   ,-1],
        [37 ,63      ,1     ,-33   ,0],
        [38 ,-58     ,-1    ,32    ,0],
        # termination
        [-1, ]
    ]

    '''amplitudes which only have constant terms; same unit as above
    {dPSI, dEPS}
    indexes which are already in ampsecul[][] are zeroed'''
    ampconst = [
        [0,0],
        [0,0],
        [46,-24],
        [11,0],
        [-3,1],
        [-3,0],
        [-2,1],
        [1,0],
        [0,0],
        [0,0],
        [0,0],
        [0,0],
        [0,0],
        [48,1],
        [-22,0],
        [0,0],
        [-15,9],
        [0,0],
        [-12,6],
        [-6,3],
        [-5,3],
        [4,-2],
        [4,-2],
        [-4,0],
        [1,0],
        [1,0],
        [-1,0],
        [1,0],
        [1,0],
        [-1,0],
        [0,0],
        [0,0],
        [0,0],
        [0,0],
        [-158,-1],
        [123,-53],
        [63,-2],
        [0,0],
        [0,0],
        [-59,26],
        [-51,27],
        [-38,16],
        [29,-1],
        [29,-12],
        [-31,13],
        [26,-1],
        [21,-10],
        [16,-8],
        [-13,7],
        [-10,5],
        [-7,0],
        [7,-3],
        [-7,3],
        [-8,3],
        [6,0],
        [6,-3],
        [-6,3],
        [-7,3],
        [6,-3],
        [-5,3],
        [5,0],
        [-5,3],
        [-4,0],
        [4,0],
        [-4,0],
        [-3,0],
        [3,0],
        [-3,1],
        [-3,1],
        [-2,1],
        [-3,1],
        [-3,1],
        [2,-1],
        [-2,1],
        [2,-1],
        [-2,1],
        [2,0],
        [2,-1],
        [1,-1],
        [-1,0],
        [1,-1],
        [-2,1],
        [-1,0],
        [1,-1],
        [-1,1],
        [-1,1],
        [1,0],
        [1,0],
        [1,-1],
        [-1,0],
        [-1,0],
        [1,0],
        [1,0],
        [-1,0],
        [1,0],
        [1,0],
        [-1,0],
        [-1,0],
        [-1,0],
        [-1,0],
        [-1,0],
        [-1,0],
        [-1,0],
        [1,0],
        [-1,0],
        [1,0]
    ]


def degrad(x):
    return x * pi / 180.


def nutation(mj):
    '''
    static double lastmj = -10000, lastdeps, lastdpsi;
    double T, T2, T3, T10;			/* jul cent since J2000 */
    double prec;				/* series precis in arc sec */
    int i, isecul;				/* index in term table */
    '''
    delcache = []
    for i in range(5):
        delcache.append([None] * (2 * Data.NUT_MAXMUL + 1))

    # lastmj = None
    # if mj == lastmj:
    #     deps = lastdeps
    #     dpsi = lastdpsi
    #     return (deps, dpsi)

    prec = 0.0
    # augment for abundance of small terms
    prec *= Data.NUT_SCALE / 10

    T = (mj - ephem.J2000) / 36525.
    T2 = T * T
    T3 = T2 * T
    T10 = T/10.
    # calculate delaunay args and place in cache
    for i in range(5):
        x = (Data.delaunay[i][0] +
             Data.delaunay[i][1] * T +
             Data.delaunay[i][2] * T2 +
             Data.delaunay[i][3] * T3)
        # convert to radians
        x /= Data.SECPERCIRC
        x -= floor(x)
        x *= 2. * pi
        # fill cache table
        for j in range(2 * Data.NUT_MAXMUL + 1):
            delcache[i][j] = (j - Data.NUT_MAXMUL) * x
    # find dpsi and deps
    lastdpsi = lastdeps = 0.

    isecul = 0
    for i in range(Data.NUT_SERIES):
        # double arg = 0., ampsin, ampcos;
        arg = 0.
        if (Data.ampconst[i][0] or Data.ampconst[i][1]):
            # take non-secular terms from simple array
            ampsin = Data.ampconst[i][0]
            ampcos = Data.ampconst[i][1]
        else:
            # secular terms from different array
            ampsin = float(Data.ampsecul[isecul][1]) + float(Data.ampsecul[isecul][2]) * T10
            ampcos = float(Data.ampsecul[isecul][3]) + float(Data.ampsecul[isecul][4]) * T10
            isecul += 1

        for j in range(5):
            arg += delcache[j][Data.NUT_MAXMUL + Data.multarg[i][j]]

        if fabs(ampsin) >= prec:
            lastdpsi += ampsin * sin(arg)

        if fabs(ampcos) >= prec:
            lastdeps += ampcos * cos(arg)
    # convert to radians.
    lastdpsi = degrad(lastdpsi/3600./Data.NUT_SCALE)
    lastdeps = degrad(lastdeps/3600./Data.NUT_SCALE)

    # lastmj = mj
    return (lastdeps, lastdpsi)


##########################################


ABERR_CONST = (20.49552 / 3600. / 180. * pi)

twopi = pi * 2.
twelfth_pi = pi / 12.
twenty_fourth_pi = pi / 24.

degree = pi / 180.
arcminute = degree / 60.
arcsecond = arcminute / 60.
half_arcsecond = arcsecond / 2.
tiny = arcsecond / 360.

_sun = ephem.Sun()    # for computing equinoxes


def get_ap_hlon(mj, nutation_dpsi=None):
    _sun.compute(mj)
    if nutation_dpsi is None:
        nutation_dpsi = nutation(mj)[1]
    ap_hlon = _sun.hlon + nutation_dpsi - ABERR_CONST + pi
    if ap_hlon < 0.0:
        ap_hlon += twopi
    elif ap_hlon > twopi:
        ap_hlon -= twopi
    return ephem.degrees(ap_hlon)


def converge(d, deg):

    def get_diff(d, deg, nutation_dpsi=None):
        diff = float(deg) * degree - get_ap_hlon(d, nutation_dpsi)
        if diff > pi:
            diff -= twopi
        elif diff < -pi:
            diff += twopi
        return diff

    # converging iterations using the get_diff() function, not fixing nutation
    for i in range(100):
        diff = get_diff(d, deg)
        if abs(diff) < ephem.degrees('0:00:01'):
            break
        d = d + 365.25 * diff / twopi  # update d using the difference

    # apply the bisection method, fixing nutation
    nutation_dpsi = nutation(d)[1]
    d0, d1 = d-ephem.degrees('0:05:00'), d+ephem.degrees('0:05:00')
    f0, f1 = get_diff(d0, deg, nutation_dpsi), get_diff(d1, deg, nutation_dpsi)
    if f0 * f1 > 0.:
        raise AssertionError("warning: f0=%f, f1=%f" % (f0, f1))

    for i in range(20):  # limits the iteration number to 20.
        dn = (d0 + d1) / 2.
        fn = get_diff(dn, deg, nutation_dpsi)
        if fn * f0 > 0.:  # the midpoint has the same sign as the left side -> select the right side
            d0 = dn
            f0 = get_diff(d0, deg, nutation_dpsi)
        elif fn * f1 > 0.:  # the midpoint has the same sign as the right side -> select the left side
            d1 = dn
            f1 = get_diff(d1, deg, nutation_dpsi)
        elif fn == 0:
            return ephem.date(dn)
        else:
            raise AssertionError("warning: impossible")

    return ephem.Date((d0*abs(f1)+d1*abs(f0))/(abs(f0) + abs(f1)))


def solar_term_finder(mj, n, reverse=False):
    def solar_term_finder_deg(mj, deg, reverse=False):
        offset = deg * degree
        d0 = ephem.Date(mj)
        motion = -twopi if reverse else twopi

        angle_to_cover = motion - (get_ap_hlon(d0) - offset) % motion
        if abs(angle_to_cover) < tiny:
            angle_to_cover = motion
        d = d0 + 365.25 * angle_to_cover / twopi
        return converge(d, deg)

    deg = ((n + 21) % 24) * 15
    return solar_term_finder_deg(mj, deg, reverse)


def solar_term_finder_adjacent(mj, divisor=30.0, remainder=15.0, reverse=False):
    d0 = ephem.Date(mj)
    if reverse:
        deg = (floor((get_ap_hlon(d0) / degree - remainder) / divisor)) * divisor + remainder
    else:
        deg = (floor((get_ap_hlon(d0) / degree - remainder) / divisor) + 1) * divisor + remainder
    angle_to_cover = deg * degree - get_ap_hlon(d0)
    d = d0 + 365.25 * angle_to_cover / twopi
    return (int((deg % 360.0) / 15.0), deg % 360.0, converge(d, deg))


def annual_solar_terms(year):
    ref = ephem.previous_winter_solstice(str(year)) + 0.01
    result = []
    for j in range(24):
        i = (j - 2) % 24
        d = pytz.utc.localize(solar_term_finder(ref, i).datetime())
        result.append((i, d))

    return result


def specified_solar_term(year, solar_term_id):
    # the reference point is between "major cold" and "start of spring"
    ref = ephem.previous_winter_solstice(str(year)) + 37.5
    return pytz.utc.localize(solar_term_finder(ref, solar_term_id).datetime())


lunarcalendar/basefestival.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals


class Festival(object):
    def __init__(self, date_func, **langs):
        '''
        :param date_func: a function which recieves a paramater called `year`,
            then return a `datetime.date()` instance.
        :param langs: language dict, such as en, zh, and more.
        '''
        self.date_func = date_func
        self.langs = {}
        # langs's format like below:
        # self.langs['en'] = [xxxx, xxxx, ...]
        # self.langs['zh'] = [xxxx, xxxx, ...]
        for key in langs:
            self.langs[key] = langs[key].split(',')

    def get_lang(self, lang):
        ''' get the first name in the target language '''
        if lang not in self.langs:
            raise NotImplemented
        return self.langs[lang][0]

    def get_lang_list(self, lang):
        ''' get the name list in the target language '''
        if lang not in self.langs:
            raise NotImplemented
        return self.langs[lang]

    def __call__(self, year):
        return self.date_func(year)


lunarcalendar/festival.py
# -*- coding: utf-8 -*-
''' all included festivals are defined here '''
from __future__ import unicode_literals
import datetime
from dateutil.easter import easter
from dateutil.rrule import rrule, MONTHLY, SU, TH
from .converter import Lunar
from .basefestival import Festival
from .solarterm import QingMing, DongZhi


# ######### Solar Festival ##########


NewYear = Festival(
        lambda year: datetime.date(year, 1, 1),
        en="New Year's Day",
        zh_hans="元旦,新年",
        )

Valentine = Festival(
        lambda year: datetime.date(year, 2, 14),
        en="Valentine's Day",
        zh_hans="情人节",
        )

WomenDay = Festival(
        lambda year: datetime.date(year, 3, 8),
        en="Women's Day",
        zh_hans="妇女节",
        )

ArborDay = Festival(
        lambda year: datetime.date(year, 3, 12),
        en="Arbor Day",
        zh_hans="植树节",
        )

ChingMing = Festival(
        lambda year: QingMing(year),
        en="Ching Ming Festival",
        zh_hans="清明节,踏青节",
        )

LabourDay = Festival(
        lambda year: datetime.date(year, 5, 1),
        en="Labour Day",
        zh_hans="劳动节",
        )

YouthDay = Festival(
        lambda year: datetime.date(year, 5, 4),
        en="Youth Day in China",
        zh_hans="青年节",
        )

NurseDay = Festival(
        lambda year: datetime.date(year, 5, 12),
        en="International Nurses Day",
        zh_hans="护士节",
        )

MotherDay = Festival(
        lambda year: rrule(MONTHLY, count=1, bymonth=5, byweekday=SU(2), dtstart=datetime.datetime(year, 5, 1))[0].date(),
        en="Mother's Day",
        zh_hans="母亲节",
        )

ChildrenDay = Festival(
        lambda year: datetime.date(year, 6, 1),
        en="Children's Day",
        zh_hans="儿童节",
        )

FatherDay = Festival(
        lambda year: rrule(MONTHLY, count=1, bymonth=6, byweekday=SU(3), dtstart=datetime.datetime(year, 6, 1))[0].date(),
        en="Father's Day",
        zh_hans="父亲节",
        )

TeacherDay = Festival(
        lambda year: datetime.date(year, 9, 10),
        en="Teacher's Day",
        zh_hans="教师节,老师节",
        )

NationDay = Festival(
        lambda year: datetime.date(year, 10, 1),
        en="National Day of the People's Republic of China",
        zh_hans="国庆节",
        )

Halloween = Festival(
        lambda year: datetime.date(year, 10, 31),
        en="Halloween",
        zh_hans="万圣夜,万圣节前夜,万鬼节",
        )

Thanksgiving = Festival(
        lambda year: rrule(MONTHLY, count=1, bymonth=11, byweekday=TH(4), dtstart=datetime.datetime(year, 11, 1))[0].date(),
        en="Thanksgiving Day",
        zh_hans="感恩节"
        )

ChristmasEve = Festival(
        lambda year: datetime.date(year, 12, 24),
        en="Christmas Eve",
        zh_hans="平安夜,圣诞夜,圣诞节前夕",
        )

ChristmasDay = Festival(
        lambda year: datetime.date(year, 12, 25),
        en="Christmas Day",
        zh_hans="圣诞节,圣诞,耶诞节,基督弥撒"
        )

Easter = Festival(
        lambda year: easter(year),
        en="Easter",
        zh_hans="复活节,主复活日",
        )

# ######### Lunar Festival ##########


LaBa = Festival(
        lambda year: Lunar(year-1, 12, 8).to_date(),
        en="LaBa Festival",
        zh_hans="腊八节",
        )

NewYearEve = Festival(
        lambda year: Lunar(year, 1, 1).to_date() - datetime.timedelta(days=1),
        en="New Year's Eve",
        zh_hans="除夕,除夕夜,大年夜,年夜,年三十,除夜,岁除,大晦日",
        )

XiaoNian = Festival(
        lambda year: Lunar(year-1, 12, 23).to_date(),
        en="XiaoNian",
        zh="小年,谢节,送灶,祭灶节,灶王节,送神",
        )

ChineseNewYear = Festival(
        lambda year: Lunar(year, 1, 1).to_date(),
        en="Chinese New Year",
        zh_hans="春节,中国新年,年结,岁首,新春,正旦,正月朔日,过新年,过年")

PoWu = Festival(
        lambda year: Lunar(year, 1, 5).to_date(),
        en="PoWu Festival",
        zh="破五节,隔开日,接财神,迎财神",
        )

Lantern = Festival(
        lambda year: Lunar(year, 1, 15).to_date(),
        en="Lantern Festival",
        zh_hans="元宵节,上元节,小正月,元夕,灯节,灯笼节",
        )

DragonHead = Festival(
        lambda year: Lunar(year, 2, 2).to_date(),
        en="Dragon Head Festival",
        zh_hans="龙抬头,龙头节,春龙节,春耕节,农事节",
        )

DragonBoat = Festival(
        lambda year: Lunar(year, 5, 5).to_date(),
        en="Dragon Boat Festival",
        zh_hans="端午节,龙舟节,端阳节,端日节,午日节,粽子节,五日节,五月节,天中节,菖蒲节",
        )

Qixi = Festival(
        lambda year: Lunar(year, 7, 7).to_date(),
        en="Qixi Festival",
        zh_hans="七夕节,七巧节,乞巧节,七姐诞,七娘生",
        )

Ghost = Festival(
        lambda year: Lunar(year, 7, 15).to_date(),
        en="Ghost Festival",
        zh_hans="中元节,鬼节,施孤,七月半,盂兰盆节",
        )

MidAutumn = Festival(
        lambda year: Lunar(year, 8, 15).to_date(),
        en="Mid-Autumn Festival",
        zh_hans="中秋节,月夕,秋节,仲秋节,八月节,八月会,追月节,玩月节,拜月节,女儿节,团圆节,赏月节",
        )

ChongYang = Festival(
        lambda year: Lunar(year, 9, 9).to_date(),
        en="Double Ninth Festival",
        zh_hans="重阳节,重九节,晒秋节,踏秋,老人节,敬老节,登山节,登高节",
        )

Hanyi = Festival(
        lambda year: Lunar(year, 10, 1).to_date(),
        en="HanYi Festival",
        zh_hans="寒衣节,十月朝,祭祖节,冥阴节,鬼头日,包袱节",
        )

DongJie = Festival(
        lambda year: DongZhi(year),
        en="Dong Festival",
        zh_hans="冬至,冬节",
        )


# ######### data ##########
festivals = [f for _, f in globals().items() if isinstance(f, Festival)]
zh_festivals = [f for f in festivals if f.get_lang('zh_hans')]
